### DISCUSSION TITLE
`nx.configuration_model` and stubs

### DISCUSSION THREAD

**Discussion by miticollo** [networkx/networkx#7112] [Q&A]
I open a question and not a bug because _maybe I could be wrong_. 

Consider the following MWE:
```python
G = nx.Graph()
G.add_edge(0,1)
G.add_edge(0,2)
G.add_edge(0,3)
G.add_edge(1,4)
G.add_edge(3,4)
degree_sequence = [d for _,d in G.degree()]
G1: nx.Graph = nx.configuration_model(degree_sequence, create_using=nx.Graph(), seed=12346)
print("G1.#edges =", G1.number_of_edges())
print("G1.#avg_degree =", np.mean([d for _,d in G1.degree()]))
print("G.#edges =", G.number_of_edges())
print("G.#degree_sequence =", degree_sequence)
print("G.#avg_degree =", np.mean(degree_sequence))

fig, axes = plt.subplots(nrows=1, ncols=2)
ax = axes.flatten()
nx.draw(G1, ax=ax[0], with_labels=True)
ax[0].set_axis_off()
ax[0].title.set_text('G1')
nx.draw(G, ax=ax[1], with_labels=True)
ax[1].set_axis_off()
ax[1].title.set_text('G')
plt.show()
```
It prints
```text
G1.#edges = 4
G1.#avg_degree = 1.6
G.#edges = 5
G.#degree_sequence = [3, 2, 1, 2, 2]
G.#avg_degree = 2.0
```
and produces
![output](https://github.com/networkx/networkx/assets/30465144/a134fe4e-c01e-480e-9d2d-8d9e3e17cc72)

**Is it correct that node 0 has degree 2 in G1 even if it has 3 stubs in G?**

**Comment by rossbar** [comment]
I think the problem stems from the fact that `nx.configuration_model` returns a multigraph by default; however, when you force the output to be cast back to a `nx.Graph` instance with `create_using=nx.Graph`, then the multiedges are lost.

If you run your above example without the `create_using` you get:

```python
>>> G1 = nx.configuration_model(deg_seq, seed=12346)
>>> G1.edges  # Notice there are two edges from node `0` to `4`
MultiEdgeView([(0, 4, 0), (0, 4, 1), (0, 3, 0), (1, 2, 0), (2, 3, 0)])
```

In this case, `G1` does indeed have the expected degree distribution:

```python
>>> G1.degree
MultiDegreeView({0: 3, 1: 1, 2: 2, 3: 2, 4: 2})
```

I'm not an expert on `configuration_model`, but this feels like a bug to me since the desired property of the output graph (i.e. degree distribution) depends on support for multiedges. From this example, I would think that `create_using=nx.Graph` should be disallowed.

**Comment by dschult** [comment]
My understanding is that the configuration model algorithm creates multiedges by design.  Otherwise it would not be able to satisfy the constraints of all the different node-degrees most of the time (indeed, the number of graphs that have a given degree sequence can be 1 much more often than my intuition says it should be -- it's pretty rare in the space of all degree sequences, but I thought it would be very very rare).  So, following the configuration model, our function returns a multigraph that matches the degree_sequence. We encourage people who want a simple graph to do exactly what you did with the `create_using` keyword argument. This results in a graph that does not exactly match the degree sequence. But it is usually close.
